### PR TITLE
Optimized damping storage and remove init kernels from GPU

### PR DESCRIPTION
**Pull request by XL64**
Remove initializations kernels from GPU and optimize the boundary condition storage to use less memory on GPU.

On Elastic FirstOrder:
2023_06_06_04_37_PM/elas3D_firstOrder/out:Umpire          DEVICE sum across ranks:  633.9 MB
To
2023_06_20_05_00_PM/elas3D_firstOrder/out:Umpire          DEVICE sum across ranks:  272.9 MB

On Elastic:
2023_06_06_04_37_PM/elas3D/out:Umpire          DEVICE sum across ranks:  494.3 MB
To
2023_06_20_05_00_PM/elas3D/out:Umpire          DEVICE sum across ranks:  137.1 MB

On acoustic:
2023_06_06_04_37_PM/acous3D/out:Umpire          DEVICE sum across ranks:  439.5 MB
to
2023_06_20_05_00_PM/acous3D/out:Umpire          DEVICE sum across ranks:   85.9 MB

On acoustic FirstOrder:
2023_06_09_01_33_PM/acous3D_firstOrder/out:Umpire          DEVICE sum across ranks:  515.2 MB
to
2023_06_20_05_00_PM/acous3D_firstOrder/out:Umpire          DEVICE sum across ranks:  157.8 MB

UPDATE:
========
Move back initialisation on GPU and free data on device after use.
Using commit acous3D/elas3D/acous3D_firstOrder/elas3D_firstOrder test cases with a 150x150x150 domain and removing I/Os:
- ACOUS3D
2023_07_25_05_10_PM/DEVELOP_acous3D/out:Umpire          DEVICE sum across ranks:    1.4 GB
2023_07_25_05_10_PM/DEVELOP_acous3D/out:total time            00:00:43.704412520
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_acous3D/out:Umpire          DEVICE sum across ranks:  935.2 MB
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_acous3D/out:total time            00:00:43.727507134
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_acous3D/out:Umpire          DEVICE sum across ranks:  778.7 MB
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_acous3D/out:total time            00:00:43.379959889

- ACOUS3D_FIRSTORDER
2023_07_25_05_10_PM/DEVELOP_acous3D_firstOrder/out:Umpire          DEVICE sum across ranks:    1.7 GB
2023_07_25_05_10_PM/DEVELOP_acous3D_firstOrder/out:total time            00:00:44.188338477
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_acous3D_firstOrder/out:Umpire          DEVICE sum across ranks:    1.2 GB
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_acous3D_firstOrder/out:total time            00:00:43.615472163
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_acous3D_firstOrder/out:Umpire          DEVICE sum across ranks:  791.5 MB
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_acous3D_firstOrder/out:total time            00:00:44.109099991

- ELAS3D
2023_07_25_05_10_PM/DEVELOP_elas3D/out:Umpire          DEVICE sum across ranks:    1.6 GB
2023_07_25_05_10_PM/DEVELOP_elas3D/out:total time            00:00:45.431493143
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_elas3D/out:Umpire          DEVICE sum across ranks:    1.3 GB
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_elas3D/out:total time            00:00:51.481615784
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_elas3D/out:Umpire          DEVICE sum across ranks:    1.0 GB
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_elas3D/out:total time            00:00:46.272910818

- ELAS3D_FIRSTORDER
2023_07_25_05_10_PM/DEVELOP_elas3D_firstOrder/out:Umpire          DEVICE sum across ranks:    2.1 GB
2023_07_25_05_10_PM/DEVELOP_elas3D_firstOrder/out:total time            00:00:51.671445793
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_elas3D_firstOrder/out:Umpire          DEVICE sum across ranks:    1.8 GB
2023_07_25_05_10_PM/DISABLE_FREEONDEVICE_elas3D_firstOrder/out:total time            00:00:51.675999238
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_elas3D_firstOrder/out:Umpire          DEVICE sum across ranks:    1.1 GB
2023_07_25_05_10_PM/ENABLE_FREEONDEVICE_elas3D_firstOrder/out:total time            00:00:45.822177176

### CODE REVIEW

File: src/coreComponents/physicsSolvers/wavePropagation/WaveSolverUtils.hpp [hunk 1]
```diff
@@ -289,8 +292,67 @@ struct WaveSolverUtils
 
 
 
-};
+/**
+ * Update pressure in one direction
+ */
+  static void UpdateP( NodeManager & nodeManager,
```
Comment (acitrain) [inline src/coreComponents/physicsSolvers/wavePropagation/WaveSolverUtils.hpp:298]: Seems ok to me! 
Just, I don't know if I would have called it "UpdateP" because when we go to elastic it is not the pressure which is updated. Maybe a name in link with dt like "TimeUpdate" or somethin like that I don't know, what do you think @sframba ? 
It is not a big deal anyway we can change that later (I can do it in my refactoring after if needed)

**Review by acitrain — APPROVED** [review]

File: src/coreComponents/physicsSolvers/wavePropagation/WaveSolverUtils.hpp (hunk 1 above)
Comment (XL64) [reply src/coreComponents/physicsSolvers/wavePropagation/WaveSolverUtils.hpp:304]: Indeed we can rename it in the PR or later.

**Comment by francoishamon** [comment]
Hello @XL64 
I tried to rebaseline your PR, but I am getting some numerical diffs:
```
Rank 0 is comparing /usr/WS2/hamon1/GEOS/integratedTests/tests/allTests/wavePropagation/acous3D_abc_smoke_01/0to200_restart_000000201/rank_0000000.hdf5 with /usr/WS2/hamon1/\
GEOS/integratedTests/tests/allTests/wavePropagation/baselines/acous3D_abc_smoke_01/0to200_restart_000000201/rank_0000000.hdf5                                                    
0: ********************************************************************************                                                                                              
0: Error: /Problem/domain/MeshBodies/mesh/meshLevels/Level0/nodeManager/pressure_np1                                                                                             
0:      Arrays of types float32 and float32 have 216 values of which 200 fail both the relative and absolute tests.                                                              
0:              Max absolute difference is at index (11,): value = -0.0009158224, base_value = 0.0                                                                               
0:              Max relative difference is at index (5,): value = -0.00045791152, base_value = 0.0                                                                               
0:      Statistics of the q values greater than 1.0 defined by absolute tolerance: N = 200                                                                                       
0:              max = 9.158224, mean = 2.6610472, std = 2.0607958                                                                                                                
0:              max is at index (11,), value = -0.0009158224, base_value = 0.0                                                                                                   
0:      Statistics of the q values greater than 1.0 defined by relative tolerance: N = 0                                                                                         
0: ******************************************************************************** 
```
so I think it is better to wait before merging your PR. I move it out of the merge queue

**Comment by sframba** [comment]
> Hello @XL64 I tried to rebaseline your PR, but I am getting some numerical diffs:
> 
> ```
> Rank 0 is comparing /usr/WS2/hamon1/GEOS/integratedTests/tests/allTests/wavePropagation/acous3D_abc_smoke_01/0to200_restart_000000201/rank_0000000.hdf5 with /usr/WS2/hamon1/\
> GEOS/integratedTests/tests/allTests/wavePropagation/baselines/acous3D_abc_smoke_01/0to200_restart_000000201/rank_0000000.hdf5                                                    
> 0: ********************************************************************************                                                                                              
> 0: Error: /Problem/domain/MeshBodies/mesh/meshLevels/Level0/nodeManager/pressure_np1                                                                                             
> 0:      Arrays of types float32 and float32 have 216 values of which 200 fail both the relative and absolute tests.                                                              
> 0:              Max absolute difference is at index (11,): value = -0.0009158224, base_value = 0.0                                                                               
> 0:              Max relative difference is at index (5,): value = -0.00045791152, base_value = 0.0                                                                               
> 0:      Statistics of the q values greater than 1.0 defined by absolute tolerance: N = 200                                                                                       
> 0:              max = 9.158224, mean = 2.6610472, std = 2.0607958                                                                                                                
> 0:              max is at index (11,), value = -0.0009158224, base_value = 0.0                                                                                                   
> 0:      Statistics of the q values greater than 1.0 defined by relative tolerance: N = 0                                                                                         
> 0: ******************************************************************************** 
> ```
> 
> so I think it is better to wait before merging your PR. I move it out of the merge queue

Hi @francoishamon , thanks for your test. @acitrain tested this branch yesterday on a real case, it seems to be working. We think that the numerical diff comes from a sensitivity of the integtraed test to implementation details: small differences in boundary conditions get amplified because the test case is extremely small. Therefore, we can integrate this PR without issues

**Comment by XL64** [comment]
linked to https://github.com/GEOS-DEV/LvArray/pull/298

**Comment by paveltomin** [comment]
finalize and merge, or close?

**Comment by acitrain** [comment]
> finalize and merge, or close?

Good question... Lot of things have change since this PR was open..
Lot of people are in holidays this week , we will discuss about this next week, I'll put the PR in draft for now

**Comment by rrsettgast** [comment]
@sframba Is this abandoned??

**Comment by acitrain** [comment]
We have done things differently (in a better way) so we can close it
